### DISCUSSION TITLE
nextjs/dynamic not supported with external packages

### DISCUSSION THREAD

**Discussion by josias-r** [vercel/next.js#14422] [Help]
I'm developing a separate npm package with nextjs components. Inside it, it's essential for the project to be able to split out components using dynamic imports.
I work with babel and use the following babelrc to compile my code of the external package:
```
{
  "presets": [
    [
      "next/babel",
      {
        "preset-env": {
          "exclude": ["@babel/proposal-dynamic-import"]
        }
      }
    ]
  ]
}
```

As you can see, I need to exclude the dynamic import rule, since otherwise, babel will compile it to "fake" dynamic imports, therefore breaking the code-splitting.

The above set-up works perfectly fine locally but it will **fail at deployment** with the following error:
```
Error: Not supported
09:47:27.258  
    at (path to file with dynamic import)
```

**Compiled example components (using the babel file from above and typescript):**

_dist/collections/TestCollection.js_

```
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _interopRequireWildcard2 = _interopRequireDefault(require("@babel/runtime/helpers/interopRequireWildcard"));

var _react = _interopRequireDefault(require("react"));

var _dynamic = _interopRequireDefault(require("next/dynamic"));

var __jsx = _react.default.createElement;
var TestComponent = (0, _dynamic.default)(function () {
  return Promise.resolve().then(function () {
    return (0, _interopRequireWildcard2.default)(require("./TestComponent"));
  });
}, {
  loadableGenerated: {
    webpack: function webpack() {
      return [require.resolveWeak("./TestComponent")];
    },
    modules: ["./TestComponent"]
  }
});

var TestCollection = function TestCollection() {
  return __jsx(_react.default.Fragment, null, __jsx("div", null, "THIS IS A TEST COLLECTION"), __jsx(TestComponent, null));
};

var _default = TestCollection;
exports.default = _default;
```

_dist/collections/TestComponent.js_
```
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var __jsx = _react.default.createElement;

var TestComponent = function TestComponent() {
  return __jsx("div", null, "THIS IS THE TEST COMPONENT");
};

var _default = TestComponent;
exports.default = _default;
```

**Comment by jamesmosier** [comment]
Hey @josias-r! I noticed you've reposted this a few times. It might be best to wait to see if others can help answer your question on your original Discussion topic, even if it is not at the top of the list.

Aside from that, can you provide a complete production of this? It would help others help you debug it. Additionally, where are you deploying? Have you tried to run `next build && next start` locally? Is the result the same? The more info you can provide the better!

**Reply by josias-r** [reply]
Hi @jamesmosier I was starting to think I'm stupid because my issue kept ending up as a discussion and I thought I might be posting it to the wrong thread 😂

I will add some additional description as soon as I find time 👍🏼  `next build && next start` work perfectly fine locally it's only a problem for deployment. I deploy directly w/ vercel.

**Reply by josias-r** [reply]
@jamesmosier I added an example. The "TestCollection" component imports the "TestComponent" using next/dynamic. This works locally during development and also using `npm run build` and `npm start`. It only fails after deployment.

Those files are inside the node_modules.

**Reply by josias-r** [reply]
@jamesmosier Can we move this back over to issues again as a bug?

**Reply by jamesmosier** [reply]
Anytime you alter the Babel config it is no longer an issue of what Next.js can or cannot handle. Unfortunately it would be an unsupported use case that you are trying to accomplish as Next provides the `next/dynamic` component for code splitting already. It also sounds more like an issue with Vercel rather than an issue with Next. 

That's just my 2 cents though 😄

**Reply by josias-r** [reply]
@jamesmosier Just to clarify. Yes, Nextjs does provide a component for that, and that is the exact component I'm using.
I altered the babel config exactly _because_ nextjs doesn't support the code splitting when it comes from an external package.

But I guess it's not a bug from nextjs but rather a complete missing feature then.
Nextjs should support external nextjs packages as a feature
Vercel just has a bug _after_ hacking the lacking nextjs feature.

But I'm not sure how a discussion will help. It's not something that can't be fixed without changes to nextjs or vercel deployments.

**Reply by thedv91** [reply]
@josias-r I'm also facing the same problem as you. Have you solved this issue yet?

**Reply by josias-r** [reply]
@thedv91 Unfortunately not, it's not supported by nextjs.
But I don't know what has changed since I last worked on that project, maybe there are new approaches that enable code-splitting an external package.

**Reply by thedv91** [reply]
@josias-r Thanks for your response. Currently, I must be using `next-transpile-modules`.
